src/VisualStudio/Core/Def/Implementation/FindReferences/Contexts/AbstractTableDataSourceFindUsagesContext.cs

```diff
@@ -277,7 +336,37 @@ protected async Task<(Guid, string projectName, SourceText)> GetGuidAndProjectNa
 
                 return new DocumentSpanEntry(
                     this, definitionBucket, documentSpan, spanKind,
-                    projectName, guid, sourceText, classifiedSpansAndHighlightSpan);
+                    projectName, guid, sourceText, classifiedSpansAndHighlightSpan, GetAggregatedCustomColumnsData(customColumnsDataOpt));
+            }
+
+            private ImmutableDictionary<string, string> GetAggregatedCustomColumnsData(ImmutableDictionary<string, ImmutableArray<string>> customColumnsDataOpt)
+            {
+                // Aggregate dictionary values to get column display values. For example, below input:
+                //
+                // {
+                //   { "Column1", {"Value1", "Value2"} },
+                //   { "Column2", {"Value3", "Value4"} }
+                // }
+                //
+                // will transform to:
+                //
+                // {
+                //   { "Column1", "Value1, Value2" },
+                //   { "Column2", "Value3, Value4" }
+                // }
+
+                if (customColumnsDataOpt == null || customColumnsDataOpt.Count == 0)
+                {
+                    return ImmutableDictionary<string, string>.Empty;
+                }
+
+                return customColumnsDataOpt.ToImmutableDictionary(
+                    keySelector: kvp => kvp.Key,
+                    elementSelector: kvp => GetCustomColumn(kvp.Key).GetDisplayStringForColumnValues(kvp.Value));
+
+                // Local functions.
+                AbstractFindUsagesCustomColumnDefinition GetCustomColumn(string columnName)
+                    => (AbstractFindUsagesCustomColumnDefinition)TableControl.ColumnDefinitionManager.GetColumnDefinition(columnName);
```

Comment:
i assume this cast is always safe?  i have'nt been able to convince myself it is.  what if some other feature/language already made this column?

Reply:
Valid point. I will filter out the key-value pairs for such cases.

---
In reply to: [222371741](https://github.com/dotnet/roslyn/pull/30155#discussion_r222371741) [](ancestors = 222371741)